src/seeds.gratitude.cpp

```diff
@@ -70,38 +92,88 @@ ACTION gratitude::acknowledge (name from, name to, string memo) {
         item.receivers.push_back(to);
     });
   }
+
+  // Updates ack stats
+  auto stitr = stats.rbegin();
+  auto round_id = stitr->round_id;
+
+  auto stitr2 = stats.find(round_id);
+  auto oldacks = stitr2->num_acks;
+  stats.modify(stitr2, _self, [&](auto& item) {
+      item.num_acks = oldacks + 1;
+  });
+
 }
 
+
 ACTION gratitude::testacks() {
   require_auth(get_self());
 
   auto actr = acks.begin();
+
   while (actr != acks.end()) {
-    calc_acks(actr->donor);
-    actr = acks.erase(actr);
+    _calc_acks(actr->donor);
+    actr++;
   }
 }
 
-ACTION gratitude::newround() {
+ACTION gratitude::calcacks(uint64_t start) {
+  require_auth(get_self());
+
+  auto actr = start == 0 ? acks.begin() : acks.find(start);
+
+  _calc_acks(actr->donor);
+
+  actr++;
+
+  // Recursion call
+  if (actr != acks.end()) {
+    action next_execution(
+      permission_level{get_self(), "active"_n},
+      get_self(),
+      "calcacks"_n,
+      std::make_tuple(actr->donor.value)
+    );
+
+    transaction tx;
+    tx.actions.emplace_back(next_execution);
+    tx.delay_sec = 1;
+    tx.send(actr->donor.value, _self);
+  } else {
+    // Cleanup on last exec
+    auto actr2 = acks.begin();
+    while (actr2 != acks.end()) {
+      actr2 = acks.erase(actr2);
+    }
+
+    // Start payout
+    action(
+      permission_level{get_self(), "active"_n},
+      get_self(),
+      "payround"_n,
+      std::make_tuple()
+    ).send();
+
+  }
+}
+
+// Called after all Acks are calculated
+ACTION gratitude::payround() {
   require_auth(get_self());
 
   auto contract_balance = eosio::token::get_balance(contracts::token, get_self(), seeds_symbol.code());
   uint64_t tot_accounts = get_size("balances.sz"_n);
   uint64_t volume = get_current_volume();
-
-  auto actr = acks.begin();
-  while (actr != acks.end()) {
-    calc_acks(actr->donor);
-    actr = acks.erase(actr);
-  }
+  float potkeep = config_get(gratz_potkp) / (float)100;
+  uint64_t usable_amount = contract_balance.amount - (contract_balance.amount * potkeep);
 
   auto bitr = balances.begin();
   while (bitr != balances.end()) {
     uint64_t my_received = bitr->received.amount;
```

Comment:
This is also a candidate where if the number of balances > 200 or so, it will fail

So best to do everything you want to do on all the accounts in the recursive (not really recursive but OK) action

Reply:
I see payround maybe a second batch action, after all acks are done